flink-ml-lib/src/main/java/org/apache/flink/ml/recommendation/swing/Swing.java

```diff
@@ -0,0 +1,444 @@
+            itemUsersMap.get(item).add(user);
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            userItemsMapState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "userItemsMapState",
+                                            Types.MAP(Types.LONG, Types.LIST(Types.LONG))));
+
+            OperatorStateUtils.getUniqueElement(userItemsMapState, "userItemsMapState")
+                    .ifPresent(
+                            x -> {
+                                userItemsMap = new HashMap<>(x.size());
+                                x.forEach(
+                                        (k, v) -> {
+                                            HashSet<Long> itemSet = new HashSet<>(v.size());
+                                            itemSet.addAll(v);
+                                            userItemsMap.put(k, itemSet);
+                                        });
+                            });
+
+            itemUsersMapState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "itemUsersMapState",
+                                            Types.MAP(Types.LONG, Types.LIST(Types.LONG))));
+
+            OperatorStateUtils.getUniqueElement(itemUsersMapState, "itemUsersMapState")
+                    .ifPresent(
+                            x -> {
+                                itemUsersMap = new HashMap<>(x.size());
+                                x.forEach(
+                                        (k, v) -> {
+                                            HashSet<Long> itemSet = new HashSet<>(v.size());
+                                            itemSet.addAll(v);
+                                            itemUsersMap.put(k, itemSet);
+                                        });
+                            });
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            Map tmpUerItemsMap = new HashMap(userItemsMap.size());
+            userItemsMap.forEach(
+                    (k, v) -> {
+                        tmpUerItemsMap.put(k, new ArrayList<>(v));
+                    });
+            userItemsMapState.update(Collections.singletonList(tmpUerItemsMap));
+
+            Map tmpItemUsersMap = new HashMap(itemUsersMap.size());
+            itemUsersMap.forEach(
+                    (k, v) -> {
+                        tmpItemUsersMap.put(k, new ArrayList<>(v));
+                    });
+            itemUsersMapState.update(Collections.singletonList(tmpItemUsersMap));
```

Comment:
Why do you need to re-construct the `itemUsersMap` and save it to state? How about saving `itemUsersMap` directly?

Reply:
Well, thanks for you advice. The problem I have is `itemUsersMap` cannot be used to update `itemUsersMapState`. The class of `itemUsersMap` is a Map<Long, HashSet<Long>> while `itemUsersMapState` is a ListState<Map<Long, List<Long>>>. So I need to construct a List<Long> object coming from  `itemUsersMap`.

Reply:
I see, thanks for the explanation.